context_header.go

```diff
@@ -38,29 +38,65 @@ type ContextWithHeaders interface {
 
 type headerCtx struct {
 	context.Context
+}
+
+// headersContainer stores the headers, and is itself stored in the context under `contextKeyHeaders`
+type headersContainer struct {
 	reqHeaders  map[string]string
 	respHeaders map[string]string
 }
 
+func (c *headerCtx) headers() *headersContainer {
+	if h, ok := c.Value(contextKeyHeaders).(*headersContainer); ok {
+		return h
+	}
+	return nil
+}
+
 // Headers gets application headers out of the context.
 func (c *headerCtx) Headers() map[string]string {
-	return c.reqHeaders
+	if h := c.headers(); h != nil {
+		return h.reqHeaders
+	}
+	return nil
 }
 
 // ResponseHeaders returns the response headers.
 func (c *headerCtx) ResponseHeaders() map[string]string {
-	return c.respHeaders
+	if h := c.headers(); h != nil {
+		return h.respHeaders
+	}
+	return nil
 }
 
 // SetResponseHeaders sets the response headers.
 func (c *headerCtx) SetResponseHeaders(headers map[string]string) {
-	c.respHeaders = headers
+	if h := c.headers(); h != nil {
+		h.respHeaders = headers
+		return
+	}
+	panic("SetResponseHeaders called on ContextWithHeaders not created via WrapWithHeaders")
 }
 
 // WrapWithHeaders returns a Context that can be used to make a call with request headers.
+// If the parent `ctx` is already an instance of ContextWithHeaders, its Headers will be
+// carried over, but the new `headers` will take precedence. ResponseHeaders are not copied.
 func WrapWithHeaders(ctx context.Context, headers map[string]string) ContextWithHeaders {
-	return &headerCtx{
-		Context:    ctx,
+	h := &headersContainer{
 		reqHeaders: headers,
 	}
+	if parent_h, ok := ctx.Value(contextKeyHeaders).(*headersContainer); ok {
+		if pl := len(parent_h.reqHeaders); pl > 0 {
+			mergedHeaders := make(map[string]string, len(headers) + pl)
+			for k, v := range parent_h.reqHeaders {
+				mergedHeaders[k] = v
+			}
+			for k, v := range headers {
+				mergedHeaders[k] = v
+			}
+			h.reqHeaders = mergedHeaders
+		}
+	}
```

Comment:
> There's no special "parent context mode" in the builder

Actually, there is, because the parent context is used to create the child context. So, for example, all `Value`s from the parent context become available on the child. But ok, I can make the act of calling `SetHeaders` trigger an "ignore parent headers" mode internally.

One thing I missed is that we should probably also inherit a span from the parent context, unless `SetExternalSpan` is called.